File: apps/local/js/HAPlaybackAdapter.js
/**
 * Home Assistant playback adapter used when AudioTube is embedded in HA.
 * It reads media_player states and sends playback commands through the HA
 * websocket API using the existing frontend session.
 */
import { defaultPlaybackState } from '../../../src/core/models.js';

const WS_URL = `${location.protocol === 'https:' ? 'wss' : 'ws'}://${location.host}/api/websocket`;

export class HAPlaybackAdapter {
  #socket;
  #nextId = 1;
  #pending = new Map();
  #selectedSpeakerId = null;
  #stateCallbacks = new Set();
  #state = defaultPlaybackState();
  #streamUrlResolver;

  constructor(streamUrlResolver) {
    this.#streamUrlResolver = streamUrlResolver;
  }

  async #connect() {
    if (this.#socket?.readyState === WebSocket.OPEN) return;

    const tokens = JSON.parse(localStorage.getItem('hassTokens') || '{}');
    if (!tokens.access_token || (tokens.expires && tokens.expires <= Date.now())) {
      window.top.location.href = '/';
      throw new Error('Home Assistant-Anmeldung abgelaufen.');
    }

    await new Promise((resolve, reject) => {
      const socket = new WebSocket(WS_URL);
      const timeout = setTimeout(() => {
        socket.close();
        reject(new Error('Home Assistant-WebSocket antwortet nicht. Bitte Home Assistant neu laden.'));
      }, 8000);
      const fail = error => {
        clearTimeout(timeout);
        socket.close();
        reject(error);
      };
      socket.onopen = () => {
        socket.send(JSON.stringify({ type: 'auth', access_token: tokens.access_token }));
      };
      socket.onmessage = event => {
        const message = JSON.parse(event.data);
        if (message.type === 'auth_ok') {
          clearTimeout(timeout);
          this.#socket = socket;
          resolve();
          return;
        }
        if (message.type === 'auth_invalid') {
          fail(new Error('Home Assistant-Anmeldung ist abgelaufen. Bitte Home Assistant neu laden.'));
          window.top.location.href = '/';
          return;
        }
        if (message.id && this.#pending.has(message.id)) {
          const pending = this.#pending.get(message.id);
          this.#pending.delete(message.id);
          if (message.success) pending.resolve(message.result);
          else pending.reject(new Error(message.error?.message || 'Home Assistant request failed.'));
        }
      };
      socket.onerror = () => fail(new Error('Home Assistant-WebSocket konnte nicht verbunden werden.'));
      socket.onclose = () => {
        this.#socket = null;
        for (const pending of this.#pending.values()) pending.reject(new Error('Home Assistant WebSocket geschlossen.'));
        this.#pending.clear();
      };
    });
  }

  async #call(type, payload = {}) {
    await this.#connect();
    const id = this.#nextId++;
    return new Promise((resolve, reject) => {
      this.#pending.set(id, { resolve, reject });
      this.#socket.send(JSON.stringify({ id, type, ...payload }));
    });
  }

  async getSpeakers() {
    const states = await this.#call('get_states');
    return states
      .filter(state => state.entity_id.startsWith('media_player.'))
      .map(state => ({
        id: state.entity_id,
        name: state.attributes.friendly_name || state.entity_id,
        type: 'speaker',
        isAvailable: !['unavailable', 'unknown'].includes(state.state),
        volume: Math.round((state.attributes.volume_level ?? 0.5) * 100),
      }));
  }

  async selectSpeaker(id) {
    const speakers = await this.getSpeakers();
    if (!speakers.some(speaker => speaker.id === id)) throw new Error(`Lautsprecher "${id}" nicht gefunden.`);
    this.#selectedSpeakerId = id;
    const speaker = speakers.find(item => item.id === id);
    this.#patchState({ volume: speaker.volume });
  }

  async play(track) {
    if (!this.#selectedSpeakerId) throw new Error('Kein Lautsprecher ausgewählt.');
    this.#patchState({ status: 'loading', currentTrack: track });
    const streamUrl = await this.#streamUrlResolver(track.id);
    await this.#call('call_service', {
      domain: 'media_player',
      service: 'play_media',
      service_data: {
        entity_id: this.#selectedSpeakerId,
        media_content_id: streamUrl,
        media_content_type: 'music',
      },
    });
    this.#patchState({ status: 'playing' });
  }

  async pause() { await this.#callService('media_pause'); this.#patchState({ status: 'paused' }); }
  async resume() { await this.#callService('media_play'); this.#patchState({ status: 'playing' }); }
  async stop() { await this.#callService('media_stop'); this.#patchState(defaultPlaybackState()); }
  async next() { await this.#callService('media_next_track'); }
  async previous() { await this.#callService('media_previous_track'); }

  async setVolume(level) {
    await this.#callService('volume_set', { volume_level: Math.max(0, Math.min(100, level)) / 100 });
    this.#patchState({ volume: level });
  }

  async #callService(service, serviceData = {}) {
    if (!this.#selectedSpeakerId) return;
    await this.#call('call_service', {
      domain: 'media_player',
      service,
      service_data: { entity_id: this.#selectedSpeakerId, ...serviceData },
    });
  }

  onStateChange(callback) { this.#stateCallbacks.add(callback); callback({ ...this.#state }); }
  destroy() { this.#socket?.close(); this.#stateCallbacks.clear(); }

  #patchState(patch) {
    this.#state = { ...this.#state, ...patch };
    for (const callback of this.#stateCallbacks) callback({ ...this.#state });
  }
}

File: apps/ha/custom_components/tube_audio_player/ha_playback_adapter.py
"""
Home Assistant Playback Adapter — Sprint 2 stub.
Controls Sonos speakers via HA media_player services.

Implements the same PlaybackAdapter interface as LocalPlaybackAdapter,
but translated to Python / HA service calls.

Status: STUB — methods are scaffolded, implementation due Sprint 2.
"""
from __future__ import annotations

import logging
from typing import Callable

from homeassistant.core import HomeAssistant, callback

_LOGGER = logging.getLogger(__name__)

# HA service names
SERVICE_PLAY_MEDIA  = "media_player.play_media"
SERVICE_MEDIA_PAUSE = "media_player.media_pause"
SERVICE_MEDIA_PLAY  = "media_player.media_play"
SERVICE_MEDIA_STOP  = "media_player.media_stop"
SERVICE_MEDIA_NEXT  = "media_player.media_next_track"
SERVICE_MEDIA_PREV  = "media_player.media_previous_track"
SERVICE_VOLUME_SET  = "media_player.volume_set"


class HAPlaybackAdapter:
    """
    Bridges the UI adapter interface to Home Assistant media_player services.

    Sprint 2 TODOs:
    - Resolve YouTube video to audio-only stream URL (via yt-dlp subprocess or
      dedicated HA service / add-on).
    - Call media_player.play_media with content_type='music' and the stream URL.
    - Subscribe to HA state changes for the selected entity and push PlaybackState
      updates back to the frontend via WebSocket.
    - Support Sonos groups via join/unjoin services.
    """

    def __init__(self, hass: HomeAssistant, stream_resolver) -> None:
        self._hass = hass
        self._stream_resolver = stream_resolver   # async (video_id: str) -> str
        self._selected_entity: str | None = None
        self._state_callbacks: list[Callable] = []

    # ── PlaybackAdapter interface ─────────────────────────────────────────────

    async def get_speakers(self) -> list[dict]:
        """Return all media_player entities that are Sonos devices."""
        states = self._hass.states.async_all("media_player")
        speakers = []
        for state in states:
            attrs = state.attributes
            # Filter to Sonos entities (platform attribute or entity_id prefix)
            # TODO Sprint 2: filter by integration / platform = sonos
            speakers.append({
                "id":          state.entity_id,
                "name":        attrs.get("friendly_name", state.entity_id),
                "type":        "group" if attrs.get("sonos_group") else "speaker",
                "isAvailable": state.state not in ("unavailable", "unknown"),
                "volume":      int((attrs.get("volume_level", 0.5)) * 100),
            })
        return speakers

    async def select_speaker(self, entity_id: str) -> None:
        self._selected_entity = entity_id

    async def play(self, track: dict) -> None:
        """Resolve audio stream and send to Sonos via play_media."""
        if not self._selected_entity:
            raise RuntimeError("Kein Lautsprecher ausgewählt.")

        stream_url = await self._stream_resolver(track["id"])

        await self._hass.services.async_call(
            "media_player",
            "play_media",
            {
                "entity_id":    self._selected_entity,
                "media_content_id":   stream_url,
                "media_content_type": "music",  # audio only — no video (F-10)
            },
        )

    async def pause(self) -> None:
        await self._call("media_pause")

    async def resume(self) -> None:
        await self._call("media_play")

    async def stop(self) -> None:
        await self._call("media_stop")

    async def next(self) -> None:
        await self._call("media_next_track")

    async def previous(self) -> None:
        await self._call("media_previous_track")

    async def set_volume(self, level: int) -> None:
        if not self._selected_entity:
            return
        await self._hass.services.async_call(
            "media_player",
            "volume_set",
            {"entity_id": self._selected_entity, "volume_level": level / 100},
        )

    def on_state_change(self, cb: Callable) -> None:
        self._state_callbacks.append(cb)

    def destroy(self) -> None:
        self._state_callbacks.clear()

    # ── Helpers ───────────────────────────────────────────────────────────────

    async def _call(self, service: str) -> None:
        if not self._selected_entity:
            return
        await self._hass.services.async_call(
            "media_player", service, {"entity_id": self._selected_entity}
        )
